osu.Game/Online/Chat/ChannelManager.cs

```diff
@@ -64,44 +65,51 @@ public class ChannelManager : PollingComponent, IChannelPostTarget
 
         private readonly IAPIProvider api;
+        private readonly NotificationsClientConnector connector;
 
         [Resolved]
         private UserLookupCache users { get; set; }
 
-        public readonly BindableBool HighPollRate = new BindableBool();
-
-        private readonly IBindable<bool> isIdle = new BindableBool();
+        private readonly IBindable<APIState> apiState = new Bindable<APIState>();
+        private bool channelsInitialised;
+        private ScheduledDelegate ackDelegate;
 
-        public ChannelManager(IAPIProvider api)
+        public ChannelManager(IAPIProvider api, NotificationsClientConnector connector)
         {
             this.api = api;
-            CurrentChannel.ValueChanged += currentChannelChanged;
-        }
+            this.connector = connector;
 
-        [BackgroundDependencyLoader(permitNulls: true)]
-        private void load(IdleTracker idleTracker)
-        {
-            HighPollRate.BindValueChanged(updatePollRate);
-            isIdle.BindValueChanged(updatePollRate, true);
-
-            if (idleTracker != null)
-                isIdle.BindTo(idleTracker.IsIdle);
+            CurrentChannel.ValueChanged += currentChannelChanged;
         }
 
-        private void updatePollRate(ValueChangedEvent<bool> valueChangedEvent)
+        [BackgroundDependencyLoader]
+        private void load()
         {
-            // Polling will eventually be replaced with websocket, but let's avoid doing these background operations as much as possible for now.
-            // The only loss will be delayed PM/message highlight notifications.
-            int millisecondsBetweenPolls = HighPollRate.Value ? 1000 : 60000;
+            connector.ChannelJoined += ch => Schedule(() => joinChannel(ch));
+            connector.NewMessages += msgs => Schedule(() => addMessages(msgs));
+            connector.PresenceReceived += () => Schedule(() =>
+            {
+                if (!channelsInitialised)
+                {
+                    channelsInitialised = true;
+                    // we want this to run after the first presence so we can see if the user is in any channels already.
+                    initializeChannels();
+                }
+            });
 
-            if (isIdle.Value)
-                millisecondsBetweenPolls *= 10;
+            connector.StartChat();
 
-            if (TimeBetweenPolls.Value != millisecondsBetweenPolls)
+            apiState.BindTo(api.State);
+            apiState.BindValueChanged(status =>
             {
-                TimeBetweenPolls.Value = millisecondsBetweenPolls;
-                Logger.Log($"Chat is now polling every {TimeBetweenPolls.Value} ms");
-            }
+                ackDelegate?.Cancel();
+
+                if (status.NewValue == APIState.Online)
+                {
+                    Scheduler.Add(ackDelegate = new ScheduledDelegate(() => api.Queue(new ChatAckRequest()), 0, 60000));
```

Comment:
Rather than doing it this way, I think it would be safer to only queue the next request when the previous one completes.

ie. change this to `Scheduler.Add(performChatAckRequest)` and schedule the next call within that new method on `Success` or `Fail`.

Reply:
I'm not sure it's safer going through events, but sure. I've made the change to the best of my ability in making it safe.